### PR TITLE
Add support for service configs to c-ares resolver.

### PR DESCRIPTION
**Pull request by markdroth**
This implements the functionality described in gRFC A2:

https://github.com/grpc/proposal/blob/master/A2-service-configs-in-dns.md

This also fixes a bug in the code from #10706, which broke the ability to actually specify an alternate DNS server.

I've done some basic manual testing of this, but there are no automated tests yet.  In order to do that, I think we'll need to figure out how to run a DNS server in the automated test framework.

### CODE REVIEW

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c [hunk 1]
```diff
@@ -152,8 +233,39 @@ static void dns_ares_on_resolved_locked(grpc_exec_ctx *exec_ctx, void *arg,
...
-    result = grpc_channel_args_copy_and_add(r->channel_args, &new_arg, 1);
+    static const char* args_to_remove[2];
+    size_t num_args_to_remove = 0;
+    grpc_arg new_args[3];
+    size_t num_args_to_add = 0;
+    new_args[num_args_to_add++] =
+        grpc_lb_addresses_create_channel_arg(r->lb_addresses);
+    grpc_service_config *service_config = NULL;
+    if (r->service_config_json != NULL) {
+      char* service_config_string =
```
Comment (y-zeng) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: We may need to free `service_config_string` after `grpc_channel_args_copy_and_add_and_remove()` is called.

**Comment by y-zeng** [comment]
This looks great! Please also update the tests that are using `grpc_dns_lookup_ares()`.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c [hunk 2]
```diff
@@ -144,6 +153,78 @@ static void dns_ares_on_retry_timer_locked(grpc_exec_ctx *exec_ctx, void *arg,
...
+  char *service_config = NULL;
+  for (grpc_json *choice = choices_json->child; choice != NULL;
+       choice = choice->next) {
+    if (choice->type != GRPC_JSON_OBJECT) {
+      gpr_log(GPR_ERROR, "cannot parse service config JSON string");
+      break;
+    }
+    grpc_json *service_config_json = NULL;
+    for (grpc_json *field = choice->child; field != NULL; field = field->next) {
+      // Check client language, if specified.
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: C++-style comments are fine.  They're legal in C99.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c (hunk 1 above)
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: Good catch!  Fixed.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c [hunk 3]
```diff
@@ -144,6 +153,78 @@ static void dns_ares_on_retry_timer_locked(grpc_exec_ctx *exec_ctx, void *arg,
...
+      if (strcmp(field->key, "clientLanguage") == 0) {
+        if (field->type != GRPC_JSON_ARRAY ||
+            !value_in_json_array(field, "c++")) {
+          service_config_json = NULL;
+          break;
+        }
+      }
+      // Check client hostname, if specified.
+      if (strcmp(field->key, "clientHostname") == 0) {
+        char hostname[HOST_NAME_MAX];
```
Comment (y-zeng) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: Looks like we don't have HOST_NAME_MAX on mac. It's recommended by posix, but not required to be defined.

Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: Yup, was already working on that.

Done.

**Comment by markdroth** [comment]
@nicolasnoble, please take a look at b826aa7 and let me know if I'm handling the portability issues the right way.  Thanks!

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c [hunk 4]
```diff
@@ -144,6 +154,77 @@ static void dns_ares_on_retry_timer_locked(grpc_exec_ctx *exec_ctx, void *arg,
...
+      // Check percentage, if specified.
+      if (strcmp(field->key, "percentage") == 0) {
+        if (field->type != GRPC_JSON_NUMBER) {
+          service_config_json = NULL;
+          break;
+        }
+        int random_pct = rand() % 100;
+        int percentage;
+        if (sscanf(field->value, "%d", &percentage) != 1 ||
+            random_pct > percentage) {
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.c]: why is there any randomness involved?

File: src/core/lib/iomgr/gethostname_fallback.c [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "src/core/lib/iomgr/port.h"
+
+#if !defined(GRPC_POSIX_HOST_NAME_MAX) && !defined(GRPC_POSIX_SYSCONF)
```
Comment (nicolasnoble) [inline src/core/lib/iomgr/gethostname_fallback.c]: The only small maintainability problem I see here potentially is if there's yet a new version of grpc_gethostname that gets added, this logic here is disjointed from adding a new entry (which resides in port.h).

Our main way to handle this is to add for instance `GRPC_GETHOSTNAME_FALLBACK` in port.h close to the other checks where people would need to add their new implementation anyway, for example where you check for duplicated `#defines`, you can also add the fallback definition.

Comment (markdroth) [inline src/core/lib/iomgr/gethostname_fallback.c]: Done.

**Pull request merged**